src/contrasts.jl

```diff
@@ -721,4 +727,30 @@ function pretty_mat(mat::AbstractMatrix; tol::Real=10*eps(eltype(mat)))
         return fracs
     end
 end
-    
+
+"""
+    struct Grouping <: StatsModels.AbstractContrasts end
+
+A placeholder type to indicate that a categorical variable is only used for
+grouping and not for contrasts.  When creating a [`CategoricalTerm`](@ref), this
+skips constructing the contrasts matrix which makes it robust to large numbers
+of levels, while still holding onto the vector of levels and constructing the
+level-to-index mapping (`invindex` field of the [`ContrastsMatrix`](@ref).).
+
+Note that calling `modelcols` on a `CategoricalTerm{Grouping}` is an error.
+This is intended only as a placeholder for other packages to build on.
+
+# Examples
+
+```julia
+julia> schema((; grp = string.(1:100_000)))
+# out-of-memory error
+
+julia> schema((; grp = string.(1:100_000)), Dict(:grp => Grouping()))
+"""
+struct Grouping <: StatsModels.AbstractContrasts
+end
+
+# return an empty matrix
+contrasts_matrix(::Grouping, baseind, n) = zeros(0,0)
+termnames(::Grouping, levels::AbstractVector, baseind::Integer) = levels
```

Comment:
Maybe also define `levels(::Grouping)`, by storing levels in the the `Grouping` struct? Couldn't that be useful for packages?

Reply:
I don't think you addressed this, did you?

Reply:
No, but since I'm ditching `Grouping` I thought it was okay :)

For the application I had in mind (random effect grouping terms for MixedModels.jl), there's really no utility in defining the order of the levels ahead of time, since it doesn't in any way affect the fitted model.  That's all that the levels stored on the contrasts are good for.

Also, this is called (IIRC) only after the `ContrastsMatrix` constructor has resolved any manually specified levels that are present on the contrasts with the levels that are present in the data, and that's what's being passed as the `levels` argument here.